### PR TITLE
ImageBuf: Only back IB with IC when passed an IC

### PR DESCRIPTION
**Pull request by lgritz**
The behavior of ImageBuf has always been that an IB referencing from a file is backed by ImageCache unless you carefully use read() with the "force" parameter set to true.  The IC can be really helpful if you only need *part* of the file, or if you have so many large IB's at once that you want IC to manage the memory for you to keep the total reasonable.

Over the years, I've started to wonder if this was the right approach for the default.  It's much more typical that IBs aren't big enough or numerous enough to gain much from the IC, and it's usually worse performance to have it IC-backed. Plus, now the properties of the ImageCache (size, behavior for nontiled images, etc.) can affect the performane of ImageBuf without users understanding why.

So this patch tries on a different approach: ImageBuf's only use an ImageCache underneath if explicitly passed a pointer to one in the IB constructor or reset() call. Otherwise, it just uses local memory owned by the IB. It's much simpler to reason about, and should be higher perf for most common cases.

I did add a global OIIO option, "imagebuf:use_imagecache" that can revert to the old behavior of having all IB's by default be backed by IC when reading from a file.

Changed oiiotool to match: By default, IB's own their local memory. Using the --cache option (which previously existed to set the size of the cache) also enables use of the cache, so it's still possible to use IC-backed IBs if the user needs it.

For the reader of this PR: What do you think? Is this a change you'd like to see? Is it something we should try to squeeze into 2.5? Or are we not sure enough about whether it's the right thing that we should not be merging it close to a major release?

### CODE REVIEW

**Comment by jessey-git** [comment]
Left two small comments/questions.

The presence of the `imagebuf:use_imagecache` is a great escape hatch if this causes unforeseen problems for folks and makes this change a lot more amenable to checking in late, at least as far as behavioral differences go.

File: src/libOpenImageIO/imagebuf.cpp [hunk 1]
```diff
@@ -936,81 +939,136 @@ ImageBufImpl::init_spec(string_view filename, int subimage, int miplevel,
...
+            round_to_multiple(m_xstride, OIIO_SIMD_MAX_SIZE_BYTES));
+        // ^^^ NB make it big enough for SIMD
+
+        // Go ahead and read any thumbnail that exists. Is that bad?
+        if (m_spec["thumbnail_width"].get<int>()
+            && m_spec["thumbnail_height"].get<int>()) {
+            m_thumbnail.reset(new ImageBuf);
+            m_has_thumbnail = input->get_thumbnail(*m_thumbnail.get(),
+                                                   subimage);
+            m_has_thumbnail = true;
```
[the remark this answers is not in the document]
Comment (lgritz) [reply src/libOpenImageIO/imagebuf.cpp:1064]: No, I didn't, must have been a cut and paste error. Will fix.

File: testsuite/python-imagebuf/src/test_imagebuf.py [hunk 2]
```diff
@@ -191,8 +191,10 @@ def ftupstr(tup) :
     b = oiio.ImageBuf ("../common/textures/grid.tx")
     b.init_spec ("../common/textures/grid.tx")
     b.read ()
-    print ("subimage:", b.subimage, " / ", b.nsubimages)
-    print ("miplevel:", b.miplevel, " / ", b.nmiplevels)
+    if b.nsubimages != 0:
```
[the remark this answers is not in the document]
Comment (lgritz) [reply testsuite/python-imagebuf/src/test_imagebuf.py:194]: Yes, if we're not going through the IC, there is not a way to know about the number of subimages and mipmaps until you read through the whole file. For some kinds of files, you can, for others, you can't. So ImageBuf.nsubimages is 0 to mean "I don't know yet at this point."

**Comment by lgritz** [comment]
Updated based on Jesse's comment. Interestingly, the typo you found was in the code before, undetected, it just moved from one place to the other in the course of this refactor. So killed another latent bug we didn't even know about, thanks to your eagle eyes.

**Comment by lgritz** [comment]
You satisfied with it now that I fixed the thumbnail typo, @jessey-git?

Anybody else have any concerns? There seemed to be wide enthusiasm for this behavior change when we discussed it at the last TSC meeting.

**Review by jessey-git — APPROVED** [review]
Yes, this seems fine from review. However, I don't have much in the way of actual ImageBuf+ImageCache code to test against "for real". Hope another reviewer might be able to fill in any gaps here.

**Pull request merged**